nucypher/policy/conditions/jwt.py

```
@@ -0,0 +1,156 @@
+                issuer=self.expected_issuer,
+            )
+        except jwt.exceptions.InvalidAlgorithmError:
+            raise JWTException(f"valid algorithms: {self._valid_jwt_algorithms}")
+        except jwt.exceptions.InvalidTokenError as e:
+            raise JWTException(e)
+
+        return payload
+
+
+class JWTCondition(AccessControlCondition):
+    """
+    A JWT condition can be satisfied by presenting a valid JWT token, which not only is
+    required to be cryptographically verifiable, but also must fulfill certain additional
+    restrictions defined in the condition.
+    """
+
+    CONDITION_TYPE = ConditionType.JWT.value
+
+    class Schema(AccessControlCondition.Schema, JWTVerificationCall.Schema):
+        condition_type = fields.Str(
+            validate=validate.Equal(ConditionType.JWT.value), required=True
+        )
+
+        @post_load
+        def make(self, data, **kwargs):
+            return JWTCondition(**data)
+
+    def __init__(
+        self,
+        jwt_token: str,
+        public_key: str,
+        condition_type: str = ConditionType.JWT.value,
+        name: Optional[str] = None,
+        expected_issuer: Optional[str] = None,
+    ):
+        try:
+            self.execution_call = JWTVerificationCall(
+                jwt_token=jwt_token,
+                public_key=public_key,
+                expected_issuer=expected_issuer,
+            )
+        except ExecutionCall.InvalidExecutionCall as e:
+            raise InvalidCondition(str(e)) from e
+
+        super().__init__(condition_type=condition_type, name=name)
+
+    @property
+    def jwt_token(self):
+        return self.execution_call.jwt_token
+
+    @property
+    def public_key(self):
+        return self.execution_call.public_key
+
+    @property
+    def expected_issuer(self):
+        return self.execution_call.expected_issuer
+
+    def verify(self, **context) -> Tuple[bool, Any]:
+        payload = self.execution_call.execute(**context)
```

Seems like at the moment we only have 2 outcomes:
- a non-None payload which causes verify to return `True`
OR
- raise an exception in the failure case eg. expired jwt, invalid issuer, invalid audience etc. (https://github.com/jpadilla/pyjwt/blob/master/jwt/exceptions.py#L9) i.e. we never actually return `False` from verify. 

Of course, clearly we want to raise for unsupported algorithm, but do we want to always raise for a failed validation instead of possibly returning (`False`, `None`)? Just wondering if, aside from unsupported algorithm, for an `InvalidTokenError` (or subsets of it) we would ever want to return (`False`, `None`) from `verify`?

---

Thinking about this more, we **need** it to return `False` instead of exceptions for cases when the JWT verification fails for non-error cases eg. invalid issuer, invalid audience, expired etc., because logical conditions (`CompoundCondition`, `SequentialCondition`, `IfThenElseCondition`) would not function properly because exceptions are considered error cases, not verification failures. So exceptions cause the logical condition to immediately stop its execution and not continue based on logical operations that can move on despite an outcome of `False`. 

For example, think of an `IfThenElseCondition` where the JWT condition is used for the `if` part. If the `if` part fails, then the `else` should be executed, but because an exception is raised for failure, the entire condition stops and fails because the exception is interpreted as an underlying error case within the overall condition.

---

Chatted with @cygnusv some more, and instead filed https://github.com/nucypher/nucypher/issues/3582, where the exception could optionally be considered as a condition failure as part of the logical condition itself.